source/lac/cuda_solver_direct.cu

```diff
@@ -0,0 +1,353 @@
+          std::get<2>(cusparse_matrix), std::get<1>(cusparse_matrix),
+          dense_matrix_dev, matrix.m());
+      AssertCusparse(cusparse_error_code);
+    }
+
+
+
+    void cusolverDngetrf_buffer_size(cusolverDnHandle_t cusolver_dn_handle, int m,
+                                     int n, float *dense_matrix_dev,
+                                     int &workspace_size)
+    {
+      cusolverStatus_t cusolver_error_code = cusolverDnSgetrf_bufferSize(
+                                               cusolver_dn_handle, m, n, dense_matrix_dev, m, &workspace_size);
+      AssertCusolver(cusolver_error_code);
+    }
+
+
+
+    void cusolverDngetrf_buffer_size(cusolverDnHandle_t cusolver_dn_handle, int m,
+                                     int n, double *dense_matrix_dev,
+                                     int &workspace_size)
+    {
+      cusolverStatus_t cusolver_error_code = cusolverDnDgetrf_bufferSize(
+                                               cusolver_dn_handle, m, n, dense_matrix_dev, m, &workspace_size);
+      AssertCusolver(cusolver_error_code);
+    }
+
+
+
+    void cusolverDngetrf(cusolverDnHandle_t cusolver_dn_handle, int m, int n,
+                         float *dense_matrix_dev, float *workspace_dev,
+                         int *pivot_dev, int *info_dev)
+    {
+      cusolverStatus_t cusolver_error_code =
+        cusolverDnSgetrf(cusolver_dn_handle, m, n, dense_matrix_dev, m,
+                         workspace_dev, pivot_dev, info_dev);
+      AssertCusolver(cusolver_error_code);
+    }
+
+
+
+    void cusolverDngetrf(cusolverDnHandle_t cusolver_dn_handle, int m, int n,
+                         double *dense_matrix_dev, double *workspace_dev,
+                         int *pivot_dev, int *info_dev)
+    {
+      cusolverStatus_t cusolver_error_code =
+        cusolverDnDgetrf(cusolver_dn_handle, m, n, dense_matrix_dev, m,
+                         workspace_dev, pivot_dev, info_dev);
+      AssertCusolver(cusolver_error_code);
+    }
+
+
+
+    void cusolverDngetrs(cusolverDnHandle_t cusolver_dn_handle, int m,
+                         float *dense_matrix_dev, int *pivot_dev, float *b,
+                         int *info_dev)
+    {
+      const int n_rhs = 1;
+      cusolverStatus_t cusolver_error_code =
+        cusolverDnSgetrs(cusolver_dn_handle, CUBLAS_OP_N, m, n_rhs,
+                         dense_matrix_dev, m, pivot_dev, b, m, info_dev);
+      AssertCusolver(cusolver_error_code);
+    }
+
+
+
+    void cusolverDngetrs(cusolverDnHandle_t cusolver_dn_handle, int m,
+                         double *dense_matrix_dev, int *pivot_dev, double *b,
+                         int *info_dev)
+    {
+      const int n_rhs = 1;
+      cusolverStatus_t cusolver_error_code =
+        cusolverDnDgetrs(cusolver_dn_handle, CUBLAS_OP_N, m, n_rhs,
+                         dense_matrix_dev, m, pivot_dev, b, m, info_dev);
+      AssertCusolver(cusolver_error_code);
```

Comment:
It's a pity that the cuda libraries (or LAPACK to begin with) are not templated and you have to list all those functions twice... :grinning: